### ISSUE TITLE
Memory leak in mqtt client

### ISSUE THREAD

**Issue by stephendeyoung** [mqttjs/MQTT.js#791]
I think I've discovered a memory leak in mqtt.js. Here is an example script:

https://gist.github.com/stephendeyoung/3180ce2ec527332e0ece6eee573e6ecb

Steps to reproduce:

1. Run script with 100 mqtt clients and using the node.js inspector: `NUMBER_OF_CLIENTS=100 node --inspect mqtt-memory-leak.js`
2. Wait for all mqtt clients to connect
3. Take a heap snapshot with the Chrome devtools
4. Wait 5-10 minutes
5. Take another heap snapshot
6. Wait another 5-10 minutes
7. Take a final heap snapshot

You should see from the snapshots that memory use is growing over time. Here is an example:

https://ufile.io/bdmme

You can even see this for just 1 mqtt client although you have to wait quite a long time before creating snapshots to see the effect.

For my particular use case I wanted to create over 1000 mqtt clients in a node.js script but I kept on running out of memory and it seems like this memory leak is causing the problem.

I am using version 2.16.0 of mqtt.js. Please let me know if I've made a mistake anywhere.

**Comment by mcollina** [comment]
@stephendeyoung unfortunately I will have no time to investigate this issue for the next few weeks. Feel free to send a pull request if you figure it out.

**Comment by aaamitsingh** [comment]
Any update on this issue?

**Comment by RangerMauve** [comment]
Nothing so far, I don't think anyone has looked into it in detail. If you have time, you could help by running a node process and using the new [inspector](https://nodejs.org/en/docs/guides/debugging-getting-started/) feature to take memory snapshots to find where the leak is happening.

**Comment by stephendeyoung** [comment]
@senthil5053 no i didn't unfortunately. Sorry I can't help.

**Comment by svogl** [comment]
I ran into the same issue and had a look around. I see memory use increasing in a minimal example that just receives (frequent) messages in an empty message handler. looking further, I suspect calling the work function with process.nextTick in lib/client.js 

https://github.com/mqttjs/MQTT.js/blob/0379bf53a8ec803b0ac3ccbe94e22467abe57213/lib/client.js#L259

directly calling work() instead results in a much flatter memory consumption line. I don't know which side-effects this would cause when you're reconnecting with many messages in the queues though.

Any ideas on this @stephendeyoung ?

Simon

**Comment by RangerMauve** [comment]
@svogl Which version of node are you using to test this?

**Comment by svogl** [comment]
I have a nodejs 10.15.0 running on my mac, on one of our ARM boards (where memory hurts), there's a node 9.8 running at the moment; same hack applied, memory consumption seems meander in a stable range

**Comment by RangerMauve** [comment]
Weird, there's nothing obvious that would cause this to happen. If you have time, do you think you could use the memory snapshot feature of the node inspector to see what exactly is being stored in memory?

As far as I can see, it should be garbage collectable once the function exits.

Another thing you could try is setting `node ---max-old-space-size=666` where `666` is the memory limit in MB before it starts garbage collecting more aggressively.

**Comment by svogl** [comment]
I created a small test set and a few inspector screenshots saved in this gist: 

https://gist.github.com/svogl/4425b31e45b64a5fbb65a83e6c7d7946

I call gc() some times before collecting dumps to compare them, there is no other traffic except 1k packets sent from the sender code.

Some of the packets seem to be coming from inspector itself, but I see many occasions of leftover  objects coming from a) copied buffer lists from bl/bl.js. and b) many process/nextTick generated small code blocks.

**Comment by RangerMauve** [comment]
Have you tried setting the `--max-old-space-size` flag? It might be that the garbage collector isn't kicking in because it doesn't think it's reached a high enough limit.

**Comment by svogl** [comment]
If you look at the gist link, I forced the gc to run with 
`node --inspect --gc-interval=100 --trace_gc --expose-gc reader.js
`
and explicitly calling gc() ; with the --trace_gc you get a line of information whenever it ran, so yes, it did run.

**Comment by RangerMauve** [comment]
This really seems like an issue with node internals rather than something MQTT.js has control over.
The reason I say this is the `newAsyncId` call in one of your screenshots that's a node internal.

Is this reproducible if you generate some junk data and process it within `process.nextTick` calls?

**Comment by svogl** [comment]
Set aside the root cause of the leak, I rewrote nextTickWork to call work only when there are packets left to process - 
```
  function nextTickWork () {
    if (packets.length) {
      process.nextTick(work)
    }
  }
```
seems that solves it for me so far and creates less objects on the fly (good for embedded systems :) )
Does this cause any adverse side effects I am not aware of?

**Comment by rodrigomvf** [comment]
When this patch wil be release?

**Comment by brettjashford** [comment]
> When this patch wil be release?

a little late, but looks like this got released in 3.0.0

**Comment by YoDaMa** [comment]
yeah gonna go ahead and close this issue.
